src/main/java/org/candlepin/model/AbstractHibernateCurator.java

```diff
@@ -958,22 +960,28 @@ public Collection<E> lockAndLoad(Iterable<? extends Serializable> ids) {
      *  A collection of IDs to use to load
      *
      * @return
-     *  A collection of locked entities matching the given values
+     *  A list of locked entities matching the given values
      */
-    protected Collection<E> lockAndLoad(Class<E> entityClass, Iterable<? extends Serializable> ids) {
+    protected List<E> lockAndLoad(Class<E> entityClass, Iterable<? extends Serializable> ids) {
+        if (ids == null) {
+            return new ArrayList<>();
+        }
+
         // Sort and de-duplicate the provided collection of IDs so we have a deterministic locking
         // order for the entities (helps avoid deadlock)
-        SortedSet<Serializable> idSet = new TreeSet<>();
-        for (Serializable id : ids) {
-            idSet.add(id);
-        }
+        List<? extends Serializable> ordered = StreamSupport.stream(ids.spliterator(), false)
+            .filter(Objects::nonNull)
+            .sorted()
+            .distinct()
+            .toList();
 
         // Fetch the entities from the DB...
-        if (idSet.size() > 0) {
+        if (ordered.size() > 0) {
             return this.currentSession()
                 .byMultipleIds(entityClass)
+                .enableOrderedReturn(false)
```

Comment:
are we sure this only affects the return order, and doesn't affect the actual lock order?

Reply:
Yes. While this is technically undocumented behavior, the method name here implies that it's just a poor documentation thing and not unintended or undefined behavior. In either case, the provided list is iterated and entities within L1 and L2 cache are resolved. Any remaining entities are then loaded from the DB at the end. 

In both "standard" implementations of the MultiIdLoader [1][2], the difference in processing is that the ordered variant keeps an extra bit of data for the destination index into the output array for the entities that have to be loaded from the DB.

The concerning part here is it's not at all clear when the locks themselves are applied. In both cases it looks like unless there is some query caching and ordering done in the background, it's possible that the intended lock order is potentially violated if there are any cache hits. That is, it looks like all the cache hits will be ordered, but then the DB locks will be ordered as a separate blob [3][4]. Depending on how interlaced they are, that could pull things out of order in a way that invites deadlock. Just one more reason to not do hard DB locks and instead try to rely on things like session isolation so we aren't fighting stacks of dependency shenanigans like this.

[1] https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/loader/ast/internal/MultiIdEntityLoaderStandard.java
[2] https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/loader/ast/internal/MultiIdEntityLoaderArrayParam.java
[3] https://github.com/hibernate/hibernate-orm/blob/2e2d2948b0fd5042c9e8a60be8a2c8e5e984767a/hibernate-core/src/main/java/org/hibernate/loader/ast/internal/MultiIdEntityLoaderStandard.java#L156
[4] https://github.com/hibernate/hibernate-orm/blob/2e2d2948b0fd5042c9e8a60be8a2c8e5e984767a/hibernate-core/src/main/java/org/hibernate/loader/ast/internal/MultiIdEntityLoaderStandard.java#L381